adapters/Vungle/VungleAdapter/GADMAdapterVungleRouter.m

```diff
@@ -25,7 +28,7 @@ @implementation GADMAdapterVungleRouter {
   NSMapTable<NSString *, id<GADMAdapterVungleDelegate>> *_initializingDelegates;
 
   /// Map table to hold the banner ad delegates with placement ID as a key.
-  NSMapTable<NSString *, id<GADMAdapterVungleDelegate>> *_bannerDelegates;
+  NSMapTable<GADMAdapterVungleBannerRequest *, NSMutableOrderedSet<id<GADMAdapterVungleDelegate>> *> *_bannerDelegates;
```

Comment:
Why is the value in this MapTable now a set? Doesn't Vungle still only support a single banner per placement ID? This makes all of the code dealing with this set more complicated than it needs to be.

Or does Vungle only support 1 banner ad at a time period (even 2 banners with unique placement IDs not supported)? Based on implementation of canRequestBannerAdForPlacementID: and the fact that the entire router class only holds 1 _bannerPlacementDI, it seems like only 1 banner is supported. If that's the case, this code can be greatly simplified by removing this _bannerDelegates altogether, replacing _bannerPlacementID to be a GADMAdapterVungleBannerRequest object, and immediately denying any requests coming in if _bannerPlacementID is not nil and doesn't match the UUID of the next request (possible refresh) coming in.

Reply:
@ericleich Yes, Vungle only supports 1 Banner per placement ID at a time even 2 Banners with unique placement ID does not support. After one Banner ad is closed(the _bannerPlacementID will be set to nil), the SDK could load and display another placement ID Banner.

But When the old Banner request object is displaying a Banner ad, the refresh Banner request object is loading a new Banner ad and also needs to be added to the _bannerDelegates, when the refresh Banner request object constructs successfully, before adding to the screen, the old Banner request object will be closed and removed from _bannerDelegates. During this period, there are two Banner request objects in _bannerDelegates. Our 6.5.2.0 Adapter updates the logic to track Banner request object lifecycle, after calling the delegate method `vungleDidCloseAdWithViewInfo:placementID:`, then the adapter could remove the old Banner request object  from _bannerDelegates.

The previous implementation is that one placementID only corresponds to one delegate, which could lead the refresh request delegate cannot be added to _bannerDelegates, after the old delegate is removed from _bannerDelegates, the _bannerDelegates is empty, so the the refresh request delegate cannot receive any callbacks. Thanks a lot. 🙂

Reply:
Understand the reasoning for the list of adapters for a single placement. But given that Vungle can currently only load 1 ad per placement, I think we can just get rid of the HashMap though and support just a single array of adapters. And per my comment above, leverage [NSPointerArray](https://developer.apple.com/documentation/foundation/nspointerarray?language=objc) with an option for NSPointerFunctionsWeakMemory to its delegate objects.

This should greatly simplify a lot of the methods. And if a new request comes and the array has a non-nil object with a different placement ID or UUID, we fail the request.

Reply:
@ericleich Hi Eric, thanks a lot for your guidance. I have a concern, if we move away from a map and use NSPointerArray with an option for NSPointerFunctionsWeakMemory to manage Banner delegates. During my debugging, I found that AdMob SDK always removes old Banner delegate and set it to nil before Vungle Adapter receives the callback method
`- (void)vungleWillCloseAdWithViewInfo:placementID:` during refreshing process,  this could lead the old Banner delegate cannot receive `- (void)willCloseAd:didDownload:` delegate method. This is why we update the logic to track Banner delegate lifecycle, after Vungle Adapter receives the callback method `- (void)vungleDidCloseAdWithViewInfo:placementID:', the adapter will remove the old Banner delegate from _bannerDelegates. Thanks a lot.  :)

Reply:
@ericleich Hi Eric, thanks a lot for your instruction. But in the coming version of Vungle SDK, multiple Banners will be supported. If we remove NSMapTable and use NSPointerArray in current version, we need to add NSMapTable back in the coming version to manage displaying multiple Banners at the same time.  😊

Reply:
Indeed, better to plan for the future then. Let's keep NSMapTable.

I'd still actually like to reopen the discussion of needing the array though. In your comment above, you mentioned it was due to losing a callback, but I don't think that callback was doing anything important and it's a sign that the publisher is no longer using that banner.

That said, would `NSMapTable<GADMAdapterVungleBannerRequest *, id<GADMAdapterVungleDelegate>>` still work?

Reply:
@ericleich Hi Eric, I agree with you.  If Google SDK is dropping the instance, it means it also dropped the ad that the adapter is responding to. So I think we don't need to track the `- (void)vungleDidCloseAdWithViewInfo:placementID: ` callback if the Banner delegate is set to nil by Google SDK . We should let Google SDK manage the Banner delegate lifecycle.  I will change the `_bannerDelegates` back to use `NSMapTable<GADMAdapterVungleBannerRequest *, id<GADMAdapterVungleDelegate>>`, thanks so much for all the help and instructions. Have a nice day. 🙂

Reply:
@ericleich Hi Eric, we changed the `_bannerDelegates` back to use `NSMapTable<GADMAdapterVungleBannerRequest *, id<GADMAdapterVungleDelegate>>`, thanks a lot for your guidance. :)